core/chains/evm/logpoller/log_poller_test.go

```diff
@@ -1,91 +1,378 @@
-	// an error without batching, this test makes sure batching is enabled.
-	for i := 0; i < 15000; i++ {
-		logs = append(logs, GenLog(th.ChainID, int64(i+1), 1, "0x3", EmitterABI.Events["Log1"].ID.Bytes(), th.EmitterAddress1))
-	}
-	require.NoError(t, th.ORM.InsertLogs(logs))
-	lgs, err := th.ORM.SelectLogsByBlockRange(1, 1)
+
+	addresses := []common.Address{th.EmitterAddress1, th.EmitterAddress2}
+	topics := []common.Hash{EmitterABI.Events["Log1"].ID, EmitterABI.Events["Log2"].ID}
+
+	err := th.LogPoller.RegisterFilter(logpoller.Filter{"convertLogs", topics, addresses})
+	require.NoError(t, err)
+
+	blk, err := th.Client.BlockByNumber(ctx, nil)
+	require.NoError(t, err)
+	require.Equal(t, big.NewInt(1), blk.Number())
+	start := blk.Time()
+
+	// There is automatically a 10s delay between each block.  To make sure it's including the correct block timestamps,
+	// we introduce irregularities by inserting two additional block delays. We can't control the block times for
+	// blocks produced by the log emitter, but we can adjust the time on empty blocks in between.  Simulated time
+	// sequence:  [ #1 ] ..(10s + delay1).. [ #2 ] ..10s.. [ #3 (LOG1) ] ..(10s + delay2).. [ #4 ] ..10s.. [ #5 (LOG2) ]
+	const delay1 = 589
+	const delay2 = 643
+	time1 := start + 20 + delay1
+	time2 := time1 + 20 + delay2
+
+	require.NoError(t, th.Client.AdjustTime(delay1*time.Second))
+	hash := th.Client.Commit()
+
+	blk, err = th.Client.BlockByHash(ctx, hash)
+	require.NoError(t, err)
+	require.Equal(t, big.NewInt(2), blk.Number())
+	assert.Equal(t, time1-10, blk.Time())
+
+	_, err = th.Emitter1.EmitLog1(th.Owner, []*big.Int{big.NewInt(1)})
+	require.NoError(t, err)
+	hash = th.Client.Commit()
+
+	blk, err = th.Client.BlockByHash(ctx, hash)
+	require.NoError(t, err)
+	require.Equal(t, big.NewInt(3), blk.Number())
+	assert.Equal(t, time1, blk.Time())
+
+	require.NoError(t, th.Client.AdjustTime(delay2*time.Second))
+	th.Client.Commit()
+	_, err = th.Emitter2.EmitLog2(th.Owner, []*big.Int{big.NewInt(2)})
+	require.NoError(t, err)
+	hash = th.Client.Commit()
+
+	blk, err = th.Client.BlockByHash(ctx, hash)
 	require.NoError(t, err)
-	// Make sure all logs are inserted
-	require.Equal(t, len(logs), len(lgs))
+	require.Equal(t, big.NewInt(5), blk.Number())
+	assert.Equal(t, time2, blk.Time())
+
+	query := ethereum.FilterQuery{
+		FromBlock: big.NewInt(2),
+		ToBlock:   big.NewInt(5),
+		Topics:    [][]common.Hash{topics},
+		Addresses: []common.Address{th.EmitterAddress1, th.EmitterAddress2}}
+
+	gethLogs, err := th.Client.FilterLogs(ctx, query)
+	require.NoError(t, err)
+	require.Len(t, gethLogs, 2)
+
+	lb, _ := th.LogPoller.LatestBlock()
+	th.PollAndSaveLogs(context.Background(), lb+1)
+	lg1, err := th.LogPoller.Logs(0, 20, EmitterABI.Events["Log1"].ID, th.EmitterAddress1)
+	require.NoError(t, err)
+	lg2, err := th.LogPoller.Logs(0, 20, EmitterABI.Events["Log2"].ID, th.EmitterAddress2)
+	require.NoError(t, err)
+
+	// Logs should have correct timestamps
```

Comment:
@reductionista I pulled this assertion to a higher level (exposed an issue in the simulated_backend_client as a side bonus) keeping it blackbox style - I believe the scan etc. assertions you had before would be covered by just a DB log read like this?

Reply:
Yeah, I guess that makes sense.  As long as it calls both `blocksFromLogs()` and `convertLogs()` internally, we shouldn't lose any testing coverage.

The only minor thing I notice is gone is the check that the address & topics are also right, but I guess we have that tested in enough other places we can just validate timestamps here.